### ISSUE TITLE
align_corners in nn.functional.grid_sample has reversed behavior

### ISSUE THREAD

**Issue by jzhangbs** [pytorch/pytorch#40153]
## 🐛 Bug

According to the [doc](https://pytorch.org/docs/stable/nn.functional.html#grid-sample), when align_corners is True, the input grid coordinates should be the position of the pixel center. And when align_corners is False, it should be the index of the pixel. e.g. if True, I need to input (0.5, 0.5) and normalized it by (0.5, 0.5)/(size/2)-1 to get the upper-left pixel. Otherwise, I need to input (0,0) and  normalized it by (0, 0)/((size-1)/2)-1

But judging from the CPU implementation [here](https://github.com/pytorch/pytorch/blob/ab26dfb44e06f1149a21304fd867a7a9ea1e22ca/aten/src/ATen/native/cpu/GridSamplerKernel.cpp#L194-L203) and [here](https://github.com/pytorch/pytorch/blob/ab26dfb44e06f1149a21304fd867a7a9ea1e22ca/aten/src/ATen/native/cpu/GridSamplerKernel.cpp#L275-L284), it unnormalize by size/2 and minus 0.5 for me in the case of align_corners=False. which is not correct. 

## To Reproduce
The following code generates a dummy array and picks the value from the original place. The output of grid_sample should be identical to the input. 

```
import torch
from torch.nn.functional import grid_sample
import numpy as np

a = np.zeros((1,2,10,10), dtype=np.float32)
for i in range(10):
    for j in range(10):
        a[0,0,i,j] = j
        a[0,1,i,j] = i
at = torch.from_numpy(a)
pos = at.permute(0,2,3,1)  # should be pixel index, used when align_corners=False

# case 1: grid is the pixel center, align_corners=False, output: 2.3842e-06
print((grid_sample(at, (pos+0.5)/10*2-1, align_corners=False)-at).abs().sum())
# case 2: grid is the pixel index, align_corners=True, output: 3.5763e-06
print((grid_sample(at, pos/9*2-1, align_corners=True)-at).abs().sum())
# case 3: grid is the pixel center, align_corners=True, output: 50.0000
print((grid_sample(at, (pos+0.5)/10*2-1, align_corners=True)-at).abs().sum())
# case 4: grid is the pixel index, align_corners=False, output: 199.4445
print((grid_sample(at, pos/9*2-1, align_corners=False)-at).abs().sum())

# test gpu kernel, result is similar
print((grid_sample(at.cuda(), (pos.cuda()+0.5)/10*2-1, align_corners=False)-at.cuda()).abs().sum())
print((grid_sample(at.cuda(), pos.cuda()/9*2-1, align_corners=True)-at.cuda()).abs().sum())
print((grid_sample(at.cuda(), (pos.cuda()+0.5)/10*2-1, align_corners=True)-at.cuda()).abs().sum())
print((grid_sample(at.cuda(), pos.cuda()/9*2-1, align_corners=False)-at.cuda()).abs().sum())
```

## Environment
pytorch=1.5

**Comment by SsnL** [comment]
it is correct, you want 
(0.5,0.5) / size * 2 - 1, not from (0, 0), because the (0, 0) is top left corner, not top left pixel center.

**Comment by ailzhang** [comment]
Closing via https://github.com/pytorch/pytorch/issues/40153#issuecomment-645638925
